### PR TITLE
Launcher: add timeout to chrome launching

### LINKED ISSUE #363: program that runs a bundled version of chromium can not stop
Here is my code. 
```
const puppeteer = require('puppeteer');


(async() => {

const browser = await puppeteer.launch({executablePath:'chrome.exe'});
const page = await browser.newPage();
await page.setViewport({width:1024,height:768});
await page.goto('https://www.google.com' ,{waitUntil: 'networkidle'});
await page.screenshot({path: 'jd.png',fullPage:true});

browser.close();
})();
````
The path of "chrome.exe" has been added to system environment variables. When I run it use the command "node test.js", it gives no output and just blocked . What's the problem?

### PR DESCRIPTION
**Pull request by aslushnikov**
This patch:
- adds 30s timeout for chromium to report the browser target url.
- adds cleanup logic for main process SIGINT signal.

Fixes #363.

### CODE REVIEW

File: lib/Launcher.js [hunk 1]
```diff
@@ -130,20 +123,47 @@ class Launcher {
...
+        stderr,
+        '',
+        'TROUBLESHOOTING: https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md',
+        '',
+      ].join('\n')));
+    }
+
+    function onTimeout() {
+      cleanup();
+      reject(new Error('Failed to connect to chrome! The only chrome revision guaranteed to work is r' + ChromiumRevision));
```
Comment (JoelEinbinder) [inline lib/Launcher.js]: Can we mention that it failed because of the timeout?

File: lib/Launcher.js [hunk 2]
```diff
@@ -130,20 +123,47 @@ class Launcher {
...
     const rl = readline.createInterface({ input: chromeProcess.stderr });
-    rl.on('line', onLine);
-    rl.once('close', () => fulfill(''));
+    let stderr = '';
+    let listeners = [
+      helper.addEventListener(rl, 'line', onLine),
+      helper.addEventListener(rl, 'close', onClose),
+      helper.addEventListener(chromeProcess, 'exit', onClose)
+    ];
+    let timeout = setTimeout(onTimeout, 30 * 1000);
```
Comment (JoelEinbinder) [inline lib/Launcher.js]: I think we should make this optional. I can imagine some crazy cloud setup where 30 seconds to connect might not be unreasonable.

File: lib/Launcher.js (hunk 1 above)
Comment (aslushnikov) [reply lib/Launcher.js]: Done

File: lib/Launcher.js (hunk 2 above)
Comment (aslushnikov) [reply lib/Launcher.js]: Done. Introduced "timeout" option to launching

**Comment by aslushnikov** [comment]
@JoelEinbinder addressed comments, please take another look!

File: docs/api.md [hunk 3]
```diff
@@ -158,6 +158,8 @@ This methods attaches Puppeteer to an existing Chromium instance.
   - `executablePath` <[string]> Path to a Chromium executable to run instead of bundled Chromium. If `executablePath` is a relative path, then it is resolved relative to [current working directory](https://nodejs.org/api/process.html#process_process_cwd).
   - `slowMo` <[number]> Slows down Puppeteer operations by the specified amount of milliseconds. Useful so that you can see what is going on.
   - `args` <[Array]<[string]>> Additional arguments to pass to the Chromium instance. List of Chromium flags can be found [here](http://peter.sh/experiments/chromium-command-line-switches/).
+  - `handleSIGINT` <[boolean]> Close chrome process on Ctrl-C. Defaults to `true`.
+  - `timeout` <[number]> Maximum time in milliseconds to wait for chrome instance to start. Defaults to `30000` (30 seconds). Pass `0` to disable timeout.
```
Comment (JoelEinbinder) [inline docs/api.md]: the Chrome

File: lib/Launcher.js [hunk 4]
```diff
@@ -127,23 +121,52 @@ class Launcher {
...
+        stderr,
+        '',
+        'TROUBLESHOOTING: https://github.com/GoogleChrome/puppeteer/blob/master/docs/troubleshooting.md',
+        '',
+      ].join('\n')));
+    }
+
+    function onTimeout() {
+      cleanup();
+      reject(new Error(`Timed out after ${timeout} ms while trying to connect to chrome! The only chrome revision guaranteed to work is r${ChromiumRevision}`));
```
Comment (JoelEinbinder) [inline lib/Launcher.js]: Chrome

**Review by JoelEinbinder — APPROVED** [review]

File: docs/api.md (hunk 3 above)
Comment (aslushnikov) [reply docs/api.md]: Done.

**Pull request merged**